coderd/httpmw/provisionerdaemon_test.go

```diff
@@ -0,0 +1,222 @@
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			routeCtx := chi.NewRouteContext()
+			r := httptest.NewRequest(http.MethodGet, "/", nil)
+			r = r.WithContext(context.WithValue(r.Context(), chi.RouteCtxKey, routeCtx))
+			res := httptest.NewRecorder()
+
+			if test.provisionerKey != "" {
+				r.Header.Set(codersdk.ProvisionerDaemonKey, test.provisionerKey)
+			}
+			if test.provisionerPSK != "" {
+				r.Header.Set(codersdk.ProvisionerDaemonPSK, test.provisionerPSK)
+			}
+
+			httpmw.ExtractProvisionerDaemonAuthenticated(test.opts)(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			})).ServeHTTP(res, r)
+
+			require.Equal(t, test.expectedStatusCode, res.Result().StatusCode)
+			if test.expectedResponseMessage != "" {
+				require.Contains(t, res.Body.String(), test.expectedResponseMessage)
+			}
+		})
+	}
+
+	t.Run("ProvisionerKey", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		defer cancel()
+		client, db, user := coderdenttest.NewWithDatabase(t, &coderdenttest.Options{
+			LicenseOptions: &coderdenttest.LicenseOptions{
+				Features: license.Features{
+					codersdk.FeatureExternalProvisionerDaemons: 1,
+				},
+			},
+		})
+		// nolint:gocritic // test
+		key, err := client.CreateProvisionerKey(ctx, user.OrganizationID, codersdk.CreateProvisionerKeyRequest{
+			Name: "dont-TEST-me",
+		})
+		require.NoError(t, err)
+
+		routeCtx := chi.NewRouteContext()
+		r := httptest.NewRequest(http.MethodGet, "/", nil)
+		r = r.WithContext(context.WithValue(r.Context(), chi.RouteCtxKey, routeCtx))
+		res := httptest.NewRecorder()
+
+		r.Header.Set(codersdk.ProvisionerDaemonKey, key.Key)
+
+		httpmw.ExtractProvisionerDaemonAuthenticated(httpmw.ExtractProvisionerAuthConfig{
+			DB:       db,
+			Optional: false,
+		})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		})).ServeHTTP(res, r)
+
+		require.Equal(t, http.StatusOK, res.Result().StatusCode)
+	})
+
+	t.Run("ProvisionerKey_NotFound", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		defer cancel()
+		client, db, user := coderdenttest.NewWithDatabase(t, &coderdenttest.Options{
+			LicenseOptions: &coderdenttest.LicenseOptions{
+				Features: license.Features{
+					codersdk.FeatureExternalProvisionerDaemons: 1,
+				},
+			},
+		})
+		// nolint:gocritic // test
+		_, err := client.CreateProvisionerKey(ctx, user.OrganizationID, codersdk.CreateProvisionerKeyRequest{
+			Name: "dont-TEST-me",
+		})
+		require.NoError(t, err)
+
+		routeCtx := chi.NewRouteContext()
+		r := httptest.NewRequest(http.MethodGet, "/", nil)
+		r = r.WithContext(context.WithValue(r.Context(), chi.RouteCtxKey, routeCtx))
+		res := httptest.NewRecorder()
+
+		r.Header.Set(codersdk.ProvisionerDaemonKey, "5Hl2Qw9kX3nM7vB4jR8pY6tA1cF0eD5uI2oL9gN3mZ4")
```

Comment:
review: Just generated a random string which should be different from the one generate in `CreateProvisionerKey`
The objective here is to use a non-existing key to have a 404-like from DB.